Add support for more operations

Robertorosmaninho opened the pull request:
This PR introduce support for: 

- IfThenElse
- For
- While
- Assign
- BitOr, BitAnd
- And, Or
- LTE, GTE, LT, GT, EQ, NEQ
- Indices property
- Min, Max
- Cast
- Case
- Comment, BlankLine

This PR also introduces 2 tests:  sum with sparse tensors and sum with dense tensors

Robertorosmaninho committed:
Update from master

Robertorosmaninho committed:
Fixing merge issues

Robertorosmaninho committed:
Merge branch 'llvm-backend' into llvm-backend

Robertorosmaninho committed:
Update README.md - Add conda environment instructions and listing dependencies

Robertorosmaninho committed:
Solving merge conflits

Robertorosmaninho committed:
Merge branch 'llvm-backend' of https://github.com/Quansight-Labs/taco into llvm-backend

Robertorosmaninho committed:
Update README.md

Robertorosmaninho committed:
Update README.md

Commit:
Merge branch 'llvm-backend' of https://github.com/Quansight-Labs/taco into HEAD

guilhermeleobas committed:
First attempt to run taco with LLVM - no jit yet

guilhermeleobas committed:
upd

guilhermeleobas committed:
upd

guilhermeleobas committed:
generate position independent bitcode

guilhermeleobas committed:
Fix function order

guilhermeleobas committed:
use std::unique_ptr with llvm structures

Commit:
add support for Indices

Robertorosmaninho committed:
Merge Quansight-Labs/llvm-jit

Robertorosmaninho committed:
Add support fot IfThenElse and fix TensorProperty::Indices

Robertorosmaninho committed:
Add support for LTE

Robertorosmaninho committed:
Add Support to BitAnd, BitOr, EQ, NEQ, GT, LT, GTE

Robertorosmaninho committed:
Fix implementation of IfThenElse

Robertorosmaninho committed:
Fix Indices and rewrite Allocate

Robertorosmaninho committed:
Add support for AND, OR operations

Robertorosmaninho committed:
Add support for Min operation

Robertorosmaninho committed:
Add support for While operation

Robertorosmaninho committed:
Add support for Max and fix support for Min

Robertorosmaninho committed:
Fix While implementation

Robertorosmaninho committed:
Add support for Assign operation

Robertorosmaninho committed:
Rewriting Comment op

Robertorosmaninho committed:
Update on Allocate support due to issue #11

Robertorosmaninho committed:
Add comment on For op implementation

Robertorosmaninho committed:
Add Support for Case opeation - Not tested yet

Robertorosmaninho committed:
Add comment on case op

Robertorosmaninho committed:
Add support for Cast operation

Robertorosmaninho committed:
Update on bb name of IfThenElse

Robertorosmaninho committed:
Deleting unused CreateMinNum

Robertorosmaninho committed:
Add new tests

Robertorosmaninho committed:
Fix implementation of AND

Robertorosmaninho committed:
Fix implementation of OR

Robertorosmaninho committed:
Fix LT,GT,LTE,GTE implementation

Robertorosmaninho committed:
Fix VarDecl Implementation

Robertorosmaninho committed:
Fix Cast Implementation

Robertorosmaninho committed:
Fix tests names

Robertorosmaninho committed:
clangFormat

Robertorosmaninho committed:
Alternative implementation for Min/Max

Robertorosmaninho committed:
Merge branch 'llvm-backend' into llvm-backend

guilhermeleobas commented on dense_sum_test.cc:
```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include "include/taco.h"
+#include "include/taco/llvm.h"
+
```
Can you move this file to `llvm-include/` folder?

Robertorosmaninho replied:
`llvm-include/` or `llvm-examples/` ?

guilhermeleobas replied:
Sorry, `llvm-examples/`

guilhermeleobas commented on sparse_sum_test.cc:
```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include "include/taco.h"
+#include "include/taco/llvm.h"
+
```
Same! Move this file to `llvm-include/` folder

guilhermeleobas replied:
`llvm-examples/`**

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "Min");
-  throw logic_error("Not Implemented for Min.");
+
+  if (op->operands.size() == 1) {
+    value = codegen(op->operands[0]);
+    return;
+  }
```
Does this case happen in any test case? If not, add a `taco_iassert (op->operands.size() == 2)`

Robertorosmaninho replied:
I couldn't generate this example, so I'll add a `taco_iassert (op->operands.size() >= 2)`

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+        (op->type.isInt()) ? Builder->CreateICmpSLT(a, b) : Builder->CreateICmpULT(a, b);
+    value = Builder->CreateSelect(icmp, a, b);
+
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      auto c = codegen(op->operands[i]);
+      icmp =
+          (op->type.isInt()) ? Builder->CreateICmpSLT(value, c) : Builder->CreateICmpULT(value, c);
+      value = Builder->CreateSelect(icmp, value, c);
+    }
+  }
```
Can `op->operands.size()` be greater than 2? If not, you can remove this code

Robertorosmaninho replied:
Yes! This is an example: [A(i) = B(i) + C(i) + D(i)](http://tensor-compiler.org/codegen.html?expr=A(i)%20=%20B(i)%20+%20C(i)%20+%20D(i)&format=A:s:0;B:s:0;C:s:0;D:s:0)

``` C
 while ((iB < pB1_end && iC < pC1_end) && iD < pD1_end) {
    int32_t iB0 = B1_crd[iB];
    int32_t iC0 = C1_crd[iC];
    int32_t iD0 = D1_crd[iD];
    int32_t i = TACO_MIN(iB0,TACO_MIN(iC0,iD0));
    if ((iB0 == i && iC0 == i) && iD0 == i) {
      A_vals[iA] = (B_vals[iB] + C_vals[iC]) + D_vals[iD];
      iA++;
    }
...
```

guilhermeleobas replied:
Notice that `TACO_MIN` only take two arguments. I don't think there's a case where `min` takes a variable number of arguments.

Robertorosmaninho replied:
Yes, but this [codegen](https://github.com/Quansight-Labs/taco/blob/ebceca71c8181031449da164b42e63ebc6be25e1/src/codegen/codegen_llvm.cpp#L350) doesn't call min again to generate `TACO_MIN(iC0,iD0)`, instead this is the first comparison and then it compares with `iB0`.

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+          (op->type.isInt()) ? Builder->CreateICmpSGT(value, c) : Builder->CreateICmpUGT(value, c);
+      value = Builder->CreateSelect(icmp, value, c);
+    }
+  }
```
Same here!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
 void CodeGen_LLVM::visit(const Eq* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "Eq");
-  throw logic_error("Not Implemented for Eq.");
+  auto* a = codegen(op->a);
+  auto* b = codegen(op->b);
+
+  if (op->type.isFloat()) {
+    if (auto* FPa = llvm::dyn_cast<llvm::ConstantFP>(a)) {
+      if (auto* FPb = llvm::dyn_cast<llvm::ConstantFP>(b)) {
+        if (FPa->isNaN() || FPb->isNaN())
+          value = Builder->CreateFCmpUEQ(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpOEQ(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
```
If neither operands are `llvm::ConstantFP`, we don't generate any code. You need to add an else branch that dispatch a floating-point comparison when both operands are not `ConstantFP`.

Robertorosmaninho replied:
Should I assume that either operand may be a QNAN then?

guilhermeleobas replied:
Yes! But you see the "problem" in the code right now? TACO only generates code iff one (or both) of the operands are `ConstantFP`

Robertorosmaninho replied:
Solved!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+  if (op->type.isFloat()) {
+    // LLVM's minnum intrinsic only does binary ops
+    value = Builder->CreateMinNum(codegen(op->operands[0]), codegen(op->operands[1]));
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      value = Builder->CreateMinNum(value, codegen(op->operands[i]));
+    }
```
Is `op->operands.size()` greater than 2 in any test case? If not, you can remove this code.

Robertorosmaninho replied:
Yes!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+
+  if (op->type.isFloat()) {
+    // LLVM's minnum intrinsic only does binary ops
+    value = Builder->CreateMinNum(codegen(op->operands[0]), codegen(op->operands[1]));
```
If you generate both `a` and `b` before this if condition, this code can be rewritten to:
```C++
auto* a = codegen(op->operands[0]);
auto* b = codegen(op->operands[1]);
if (op->type.isFloat()) {
    value = Builder->CreateMinNum(a, b);
} else if (op->type.isInt()) {
    ...
}
```

Robertorosmaninho replied:
ok!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+    }
+  } else {
+    value = Builder->CreateICmpEQ(a, b);
+  }
```
Change the else statement to check if `op->type.isInt() || op->type.isUInt()` is true
```suggestion
  } else if (op->type.isInt() || op->type.isUInt()) {
    value = Builder->CreateICmpEQ(a, b);
  } else {
     // raise an error here
  }
```

Robertorosmaninho replied:
Done!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+        else
+          value = Builder->CreateFCmpONE(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
```
Same here! You don't generate any code if `a` and `b` are not `ConstantFP`.

Robertorosmaninho replied:
Done!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+    }
+  } else {
+    value = Builder->CreateICmpNE(a, b);
+  }
```
Same here! Ensure `op` type is either `Int` or `UInt`

Robertorosmaninho replied:
Done!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+    value = Builder->CreateICmpUGT(a, b);
+  } else {
+    value = Builder->CreateICmpSGT(a, b);
+  }
```
Same

Robertorosmaninho replied:
Done!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+    value = Builder->CreateICmpULT(a, b);
+  } else {
+    value = Builder->CreateICmpSLT(a, b);
+  }
```
Same

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+    value = Builder->CreateICmpUGE(a, b);
+  } else {
+    value = Builder->CreateICmpSGE(a, b);
+  }
```
Same

Robertorosmaninho replied:
Done!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+    value = Builder->CreateICmpULE(a, b);
+  } else {
+    value = Builder->CreateICmpSLE(a, b);
+  }
```
Same

Robertorosmaninho replied:
Done!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+    value = Builder->CreateICmpSLE(a, b);
+  }
 }
 
 void CodeGen_LLVM::visit(const And* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "And");
-  throw logic_error("Not Implemented for And.");
+  auto* actual_bb = Builder->GetInsertBlock();
+
+  // Create the BasicBlocks
+  auto* true_bb = llvm::BasicBlock::Create(*this->Context, "true_and_bb", this->Func);
+  auto* end_bb = llvm::BasicBlock::Create(*this->Context, "end_and_bb", this->Func);
+
+  // Generate first condition
+  auto a = codegen(op->a);
+  Builder->CreateCondBr(a, true_bb, end_bb);
+
+  // If true -> Generate sencond condition
+  Builder->SetInsertPoint(true_bb);
+  auto b = codegen(op->b);
+  Builder->CreateBr(end_bb);
+
+  // Create the PHI node on the end_bb to get the result from a && b
+  Builder->SetInsertPoint(end_bb);
+  auto phi = Builder->CreatePHI(get_int_type(1, *this->Context), 2);
+  phi->addIncoming(a, actual_bb);
+  phi->addIncoming(b, true_bb);
+  value = phi;
```
Can you implement this instruction as a function call? For instance, create a function in the module that implements a `logical and`, let's call this function `logical_and`. Whenever you emit code for the `And*` opcode, you emit a call to the `logical_and` function.

guilhermeleobas replied:
Ignore this comment. Doesn't make sense to create a function for this opcode.

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+  phi->addIncoming(b, true_bb);
+  value = phi;
 }
 
 void CodeGen_LLVM::visit(const Or* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "Or");
-  throw logic_error("Not Implemented for Or.");
+  auto* actual_bb = Builder->GetInsertBlock();
+
+  // Create the BasicBlocks
+  auto* false_bb = llvm::BasicBlock::Create(*this->Context, "false_or_bb", this->Func);
+  auto* end_bb = llvm::BasicBlock::Create(*this->Context, "end_or_bb", this->Func);
+
+  // Generate first condition
+  auto a = codegen(op->a);
+  Builder->CreateCondBr(a, false_bb, end_bb);
+
+  // If false -> Generate the sencond condition to test
+  Builder->SetInsertPoint(false_bb);
+  auto b = codegen(op->b);
+  Builder->CreateBr(end_bb);
+
+  // Create the PHI node on the end_bb to get the result from a || b
+  Builder->SetInsertPoint(end_bb);
+  auto phi = Builder->CreatePHI(get_int_type(1, *this->Context), 2);
+  phi->addIncoming(a, actual_bb);
+  phi->addIncoming(b, false_bb);
+  value = phi;
```
The same for `logical_or`. Given that both functions are sort of similar, I think you can reuse almost the same code to implement both functions.

guilhermeleobas replied:
Ignore this comment. Doesn't make sense to create a function for this opcode.

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -390,12 +601,49 @@ void CodeGen_LLVM::visit(const Call* op) {
+}
+}  // anonymous namespace
+
+void CodeGen_LLVM::visit(const Case* op) {  // Not tested
```
Given we don't have any tests for this opcode, let's postpone it to another PR. Just comment the call to `codegen(caseToIfThenElse(...))` and throw an error.

Robertorosmaninho replied:
Actually, this call is essential to nested IfThenElse!
This is not the `case` used for Switch, this is used to generate a series of conditionals like [here](https://github.com/Robertorosmaninho/taco/blob/4afa91a6667be480ee05de2f8ee1bb746d909b36/include/taco/ir/ir.h#L574)!

guilhermeleobas replied:
Ok! But do we have any test the exercise the code we generate?

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -484,7 +732,31 @@ void CodeGen_LLVM::visit(const For* op) {
+  Builder->SetInsertPoint(loop);
+  codegen(op->contents);
+
+  // create unconditional branch to check
```
```suggestion
  // create unconditional branch to header
```

Robertorosmaninho replied:
Ok!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -484,7 +732,31 @@ void CodeGen_LLVM::visit(const For* op) {
+  // create unconditional branch to check
+  Builder->CreateBr(header);
+
+  // seet the insert point for the exit loop
```
```suggestion
  // set the insert point for the exit loop
```

Robertorosmaninho replied:
ok!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -622,23 +904,37 @@ void CodeGen_LLVM::visit(const Yield* op) {
-    ret->setName("calloc.ret");
-  }
+  auto _ = CodeGen_LLVM::IndentHelper(this, "Warning: Missing Allocate implementation");
+  /*
+    auto _ = CodeGen_LLVM::IndentHelper(this, "Allocate");
+    auto voidptr = get_void_ptr_type(*this->Context);
+    auto i64 = get_int_type(64, *this->Context);
+    auto i32p = get_int_ptr_type(32, *this->Context);
+
+    auto alloca = this->Builder->CreateAlloca(i32p);
+  //  auto var = codegen(op->var);
+    auto num_elements = codegen(op->num_elements);
+    auto sext_num_elements = this->Builder->CreateSExt(num_elements, i64);
+
+    if (op->is_realloc) {
+      auto size = this->Builder->CreateMul(
+          sext_num_elements, get_int_constant(64, 4, *this->Context), "realloc.size");
+
+      auto ret = emitExternalCall("realloc", voidptr, {i64}, {size});
+      ret->setName("realloc.ret");
+
+      auto bitcast = this->Builder->CreateBitCast(ret, i32p);
+      this->Builder->CreateStore(bitcast, alloca);
+    } else {
+      auto size = this->Builder->CreateMul(
+        sext_num_elements, get_int_constant(64, 4, *this->Context));
+
+     auto call = emitExternalCall("malloc", voidptr, {i64}, {size});
+     call->setName("calloc.ret");
+
+     auto bitcast = this->Builder->CreateBitCast(call, i32p);
+     this->Builder->CreateStore(bitcast, alloca);
+    }*/
```
Let's do this in a later PR. Can you emit a warning here saying we're not emitting code for `Allocate`

Robertorosmaninho replied:
Should I add a `throw logic_error("Not Implemented for Allocate.");` after `auto _ = CodeGen_LLVM::IndentHelper(this, "Warning: Missing Allocate implementation");` ?

guilhermeleobas replied:
No, use `taco_uwarning` for the warning message

```c++
 taco_uwarning << "Missing LLVM implementation for Allocate opcode" << std::endl;
```

Don't throw a logic_error

Robertorosmaninho replied:
`taco_uwarning` stops the code generation. Are ther any other instruction that issue a warning without interrupting the code generation process?

guilhermeleobas replied:
Apply the following diff

```diff
diff --git a/src/error.cpp b/src/error.cpp
index 9be355db..90292a45 100644
--- a/src/error.cpp
+++ b/src/error.cpp
@@ -62,7 +62,8 @@ ErrorReport::ErrorReport(const char *file, const char *func, int line,
 void ErrorReport::explodeWithException() {
   TacoException e = TacoException(msg->str());
   delete msg;
-  throw e;
+  if (!warning)
+    throw e;
 }

 }
```

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -648,11 +944,11 @@ void CodeGen_LLVM::visit(const Free* op) {
 
 void CodeGen_LLVM::visit(const BlankLine* op) {
-  auto _ = CodeGen_LLVM::IndentHelper(this, "BlankLine");
+  // no-op, do nothing
```
```suggestion
  auto _ = CodeGen_LLVM::IndentHelper(this, "BlankLine");
  // no-op, do nothing
```

Robertorosmaninho replied:
ok!

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -715,7 +1012,19 @@ void CodeGen_LLVM::visit(const GetProperty* op) {
     case TensorProperty::ModeTypes:
-    case TensorProperty::Indices:
+    case TensorProperty::Indices: {
+      auto* mode = get_int_constant(32, op->mode, *this->Context);
+      auto* index = get_int_constant(32, op->index, *this->Context);
+
+      auto* load1 = this->Builder->CreateLoad(
+          this->Builder->CreateStructGEP(tensor, (int) TensorProperty::Indices));  // i8***
+      auto* load2 =
+          this->Builder->CreateLoad(this->Builder->CreateInBoundsGEP(load1, mode));  // i8**
+      auto* load3 =
+          this->Builder->CreateLoad(this->Builder->CreateInBoundsGEP(load2, index));  // i8*
+      value = this->Builder->CreateBitCast(load3, i32p);
+      break;
```
Let's move this to a function. This might make things easier to debug

Robertorosmaninho replied:
ok!

guilhermeleobas reviewed (changes requested):
Hi @Robertorosmaninho, your PR looks good. I can confirm both tests are running fine, the segfault you see after the code is executed is not related to your PR, and it was introduced in #10. 

I've made an initial review. Let's move the implementation of `logical_and` and `logical_or` to a function and emit a function call to it. Also, I guess the implementation for the comparison opcodes can be merged into a single function.

Let me know if you have any questions.

Robertorosmaninho committed:
Fix Max/Min implementation

Robertorosmaninho committed:
Fix comments

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +333,269 @@ void CodeGen_LLVM::visit(const Rem* op) {
+  auto a = codegen(op->a);
+  Builder->CreateCondBr(a, false_bb, end_bb);
+
+  // If false -> Generate the sencond condition to test
```
```suggestion
  // If false -> Generate the second condition to test
```

Robertorosmaninho replied:
ok!

Robertorosmaninho committed:
Moving tests to llvm-examples

Robertorosmaninho committed:
Fix lib's path on llvm-examples

Robertorosmaninho committed:
Build function to getOrInsert Indices

Robertorosmaninho committed:
Add IndentHelper for BlankLine

Robertorosmaninho committed:
Fix comments and Add Builder to LTE, GTE, LT, GT, NEQ and EQ
These builder guarantee that even if one of the operands
doesn't match with llvm::ConstantFP the code still is generated!

Robertorosmaninho committed:
Fix And implementation

Robertorosmaninho committed:
Fix Or Implementation

Robertorosmaninho committed:
Refactoring max and min to use "a" and "b"

Robertorosmaninho committed:
Fix typo

Robertorosmaninho committed:
Add iassert to guarantee that ICmpNE is used by int of uint

guilhermeleobas commented on src/codegen/codegen_llvm.h:
```diff
@@ -45,13 +45,15 @@ class CodeGen_LLVM : public CodeGen {
   void pushSymbol(const std::string& name, llvm::Value* v);
   void removeSymbol(const std::string& name);
   llvm::Value* getSymbol(const std::string& name);
+  bool containsSymbol(const std::string& name);
```
Add the following
```suggestion
  bool containsSymbol(const std::string& name);
  llvm::Function* getOrInsertGetIndicesFunction();
```

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +372,299 @@ void CodeGen_LLVM::visit(const Rem* op) {
+      value = Builder->CreateFCmpUEQ(a, b);
+    }
+  } else {
+    taco_iassert(op->type.isInt() || op->type.isUInt()) << "op->type is " << op->type;
```
`op->type` can be boolean here as well
```suggestion
    taco_iassert(op->type.isInt() || op->type.isUInt() || op->type.isBool()) << "op->type is " << op->type;
```

guilhermeleobas commented on src/codegen/codegen_llvm.cpp:
```diff
@@ -320,67 +372,299 @@ void CodeGen_LLVM::visit(const Rem* op) {
+      value = Builder->CreateFCmpUNE(a, b);
+    }
+  } else {
+    taco_iassert(op->type.isInt() || op->type.isUInt()) << "op->type is " << op->type;
```
```suggestion
    taco_iassert(op->type.isInt() || op->type.isUInt() || op->type.isBool()) << "op->type is " << op->type;
```

guilhermeleobas reviewed (changes requested):
Hi Roberto!

The PR looks good. After these fixes, we can merge it. Thanks for all the work you've been doing.

Robertorosmaninho committed:
Add getOrInsertGetIndicesFunction on codegen_llvm.h

Robertorosmaninho commented:
https://github.com/Quansight-Labs/taco/pull/12#discussion_r671435995 Well, this function doesn't create any new code just call IfThenElse for IF-ELSEIF-ELSE cases. In this example [A(i) = B(i) + C(i)](http://tensor-compiler.org/codegen.html?expr=A(i)%20=%20B(i)%20+%20C(i)&format=A:s:0;B:s:0;C:s:0) this function is called one time for `assemble` and another for `compute` to generate code for 
```
if (iB0 == i && iC0 == i) {
      A_vals[iA] = B_vals[iB] + C_vals[iC];
      iA++;
    }
    else if (iB0 == i) {
      A_vals[iA] = B_vals[iB];
      iA++;
    }
    else {
      A_vals[iA] = C_vals[iC];
      iA++;
    }
```
If the sum of the tensor has more operands, then more `else if` are needed and this function is called many times!
LLVM IR for A(i) = B(i) + C(i) with debug output (look for "Case"): https://pastebin.com/TaDmbjDj - Example: https://pastebin.com/b56gy5NV
LLVM IR for A(i) = B(i) + C(i) + D(i) with debug output (look for "Case"): [gist](https://gist.github.com/Robertorosmaninho/b9a20a56225bbfa0bf839ef6289bec08) - Example: https://pastebin.com/tND6Q8iw

Robertorosmaninho committed:
Add Waning in Allocate and Fix NEQ, EQ, LT

Robertorosmaninho committed:
clangFormat and only trows an exception if it isn't a warning

Robertorosmaninho committed:
clangFormat, deleting duplicated example and add new example

Robertorosmaninho committed:
Fix sparse_sum_test.cc example

Robertorosmaninho committed:
Fix dense_sum_test.cc example

guilhermeleobas commented on src/error.cpp:
```diff
@@ -13,12 +13,21 @@ const char* TacoException::what() const noexcept {
-                         Kind kind, bool warning)
-    : msg(NULL), file(file), func(func), line(line), condition(condition),
-      conditionString(conditionString), kind(kind), warning(warning) {
+ErrorReport::ErrorReport(const char* file,
+                         const char* func,
+                         int line,
+                         bool condition,
+                         const char* conditionString,
+                         Kind kind,
+                         bool warning)
+    : msg(NULL)
+    , file(file)
+    , func(func)
+    , line(line)
+    , condition(condition)
+    , conditionString(conditionString)
+    , kind(kind)
+    , warning(warning) {
```
Can you undo any change related to clang-format in this file?

Robertorosmaninho replied:
Yes!

guilhermeleobas reviewed (changes requested):
Hi @Robertorosmaninho, your PR looks almost ready to be merged. Just a small request and we can merge it.

Robertorosmaninho committed:
Undo error.cpp clangFormat

Robertorosmaninho committed:
only trows an exception if it isn't a warning

guilhermeleobas commented:
Thanks, @Robertorosmaninho
